### PR TITLE
[Examples] CustomCollectionView in Swift3

### PR DESCRIPTION
**Pull request by rajeevwhiz**
Using ASDK 2.0 and Swift3. CustomCollectionView uses MosaicCollectionViewLayout to keep aspect ratio of images intact and tile them perfectly.

### CODE REVIEW

**Comment by hannahmbanana** [comment]
@rajeevwhiz - thanks again for contributing this!

I have a few comments:
- remove the top level "Sample" folder. It should have the same structure as the `LayoutSpecExamples-Swift` example project.
- remove the `Sample.xcworkspace` file

**Comment by MarvinNazari** [comment]
shouldn't `asyncDataSource` and  `asyncDelegate` be deprecated by now? or am i missing something?

File: examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    
+    let layout = MosaicCollectionViewLayout.init()
+    layout.numberOfColumns = 3;
+    layout.headerHeight = 44;
+    layout.delegate = self
+    _layoutInspector = MosaicCollectionViewLayoutInspector.init()
+    
+    _collectionNode = ASCollectionNode.init(frame: CGRect.zero, collectionViewLayout: layout)
+    _collectionNode?.view.asyncDataSource = self;
```
Comment (hannahmbanana) [inline examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift]: change to _collectionNode?.dataSource = self;

File: examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    _collectionNode?.view.asyncDelegate = self;
```
Comment (hannahmbanana) [inline examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift]: change to _collectionNode?.delegate = self;

File: examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    _collectionNode?.view.layoutInspector = _layoutInspector
+    _collectionNode?.backgroundColor = UIColor.white
+    _collectionNode?.view.isScrollEnabled = true
+    _collectionNode?.registerSupplementaryNode(ofKind: UICollectionElementKindSectionHeader)
+  }
+  
+  deinit {
+    _collectionNode?.view.asyncDataSource = nil;
```
Comment (hannahmbanana) [inline examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift]: set these on the node direction too

File: examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    self.view.addSubnode(_collectionNode!)
+  }
+  
+  override func viewWillLayoutSubviews() {
+    _collectionNode?.frame = self.view.bounds;
+  }
+  
+  
+  // MARK: ASCollectionView data source
+  func collectionView(_ collectionView: ASCollectionView, nodeBlockForItemAt indexPath: IndexPath) -> ASCellNodeBlock {
```
Comment (hannahmbanana) [inline examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift]: use the collectionNode's method instead

File: examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      NSFontAttributeName: UIFont.preferredFont(forTextStyle: UIFontTextStyle.headline),
+      NSForegroundColorAttributeName: UIColor.gray
+    ]
+    let textInsets = UIEdgeInsets.init(top: 11, left: 0, bottom: 11, right: 0)
+    let textCellNode = ASTextCellNode.init(attributes: textAttributes as! [AnyHashable : Any], insets: textInsets)
+    textCellNode.text = String.init(format: "Section %zd", indexPath.section + 1)
+    return textCellNode;
+  }
+  
+  func numberOfSections(in collectionView: UICollectionView) -> Int {
```
Comment (hannahmbanana) [inline examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift]: use the collectionNode method

File: examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    return _sections!.count
+  }
+  
+  func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
```
Comment (hannahmbanana) [inline examples/CustomCollectionView-Swift/Sample/Sample/ViewController.swift]: use the collectionNode method

**Comment by hannahmbanana** [comment]
@rajeevwhiz - let me know if I can help answer any questions about the example project folder structure. :)

**Comment by rajeevwhiz** [comment]
Fixed all issues and suggestions..

**Comment by hannahmbanana** [comment]
@rajeevwhiz - it's almost ready, just 2 last things:

1. Remove the top level "Sample" folder (move all files up a level in the folder structure). 

2. Remove the Sample.xcworkspace file

Thanks!

**Comment by appleguy** [comment]
@rajeevwhiz thanks for working on this!

**Comment by rajeevwhiz** [comment]
Removed top-level Sample folder and Sample.xcworkspace file. Please confirm if everything looks good now.

File: examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    
+    for section in 0 ..< numberOfSections {
+      let numberOfItems = self.collectionView?.numberOfItems(inSection: section)
+      
+      top += _sectionInset.top
+      
+      if (headerHeight > 0) {
+        let headerSize: CGSize = self._headerSizeForSection(section: section)
+        
+        let attributes: UICollectionViewLayoutAttributes = UICollectionViewLayoutAttributes.init(forSupplementaryViewOfKind: UICollectionElementKindSectionHeader, with: NSIndexPath.init(row: 0, section: section) as IndexPath)
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift]: You don't need `let attributes: UICollectionViewLayoutAttributes =`, just use `let attributes =`, the type will be automatically inferred by the equality.
You don't need the `.init(...)` you can just use `(...)` for both the `UICollectionViewLayoutAttributes`and the `IndexPath`

File: examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    super.prepare()
+    
+    _itemAttributes = []
+    _allAttributes = []
+    _headerAttributes = []
+    _columnHeights = []
+    
+    var top: CGFloat = 0
+    
+    let numberOfSections: NSInteger = self.collectionView!.numberOfSections
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift]: You can wrap this with a `guard let collectionView = self.collectionView else { return }` and use `collectionView`rather then all the unwrapping done in this function.

File: examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+class MosaicCollectionViewLayout: UICollectionViewFlowLayout {
+  var numberOfColumns: Int
+  var columnSpacing: CGFloat
+  var _sectionInset: UIEdgeInsets
+  var interItemSpacing: UIEdgeInsets
+  var headerHeight: CGFloat
+  var _columnHeights: [[CGFloat]]?
+  var _itemAttributes: [[UICollectionViewLayoutAttributes]]?
+  var _headerAttributes: [UICollectionViewLayoutAttributes]?
+  var _allAttributes: [UICollectionViewLayoutAttributes]?
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift]: Those attributes could be initialized as empty arrays, example:
`var _allAttributes = [UICollectionViewLayoutAttributes]()`
This way you don't need to do any optional unwrapping later on. I didn't see a reason why these would/should be nil

File: examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+      let numberOfItems = self.collectionView?.numberOfItems(inSection: section)
+      
+      top += _sectionInset.top
+      
+      if (headerHeight > 0) {
+        let headerSize: CGSize = self._headerSizeForSection(section: section)
+        
+        let attributes: UICollectionViewLayoutAttributes = UICollectionViewLayoutAttributes.init(forSupplementaryViewOfKind: UICollectionElementKindSectionHeader, with: NSIndexPath.init(row: 0, section: section) as IndexPath)
+        
+        attributes.frame = CGRect.init(x: _sectionInset.left, y: top, width: headerSize.width, height: headerSize.height)
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift]: You don't need the `.init(...)` you can just use `(...)`

File: examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+      _columnHeights?.append([]) //Adding new Section
+      for _ in 0 ..< self.numberOfColumns {
+        self._columnHeights?[section].append(top)
+      }
+      
+      let columnWidth: CGFloat = self._columnWidthForSection(section: section)
+      _itemAttributes?.append([])
+      for idx in 0 ..< numberOfItems! {
+        let columnIndex: Int = self._shortestColumnIndexInSection(section: section)
+        let indexPath = IndexPath.init(item: idx, section: section)
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift]: You don't need the .init(...) you can just use (...)

File: examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+        
+        let itemSize: CGSize = self._itemSizeAtIndexPath(indexPath: indexPath);
+        let xOffset: CGFloat = _sectionInset.left + (columnWidth + columnSpacing) * CGFloat(columnIndex)
+        let yOffset: CGFloat = _columnHeights![section][columnIndex]
+        
+        let attributes: UICollectionViewLayoutAttributes = UICollectionViewLayoutAttributes.init(forCellWith: indexPath)
+        
+        attributes.frame = CGRect.init(x: xOffset, y: yOffset, width: itemSize.width, height: itemSize.height)
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift]: You don't need the .init(...) you can just use (...)

File: examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+      _itemAttributes?.append([])
+      for idx in 0 ..< numberOfItems! {
+        let columnIndex: Int = self._shortestColumnIndexInSection(section: section)
+        let indexPath = IndexPath.init(item: idx, section: section)
+        
+        let itemSize: CGSize = self._itemSizeAtIndexPath(indexPath: indexPath);
+        let xOffset: CGFloat = _sectionInset.left + (columnWidth + columnSpacing) * CGFloat(columnIndex)
+        let yOffset: CGFloat = _columnHeights![section][columnIndex]
+        
+        let attributes: UICollectionViewLayoutAttributes = UICollectionViewLayoutAttributes.init(forCellWith: indexPath)
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift]: You don't need `let attributes: UICollectionViewLayoutAttributes =`, just use `let attributes =`, the type will be automatically inferred by the equality.
You don't need the .init(...) you can just use (...)

File: examples/CustomCollectionView-Swift/Sample/ViewController.swift [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  let kNumberOfImages: UInt = 14
+  
+  required init?(coder aDecoder: NSCoder) {
+    super.init(coder: aDecoder)
+    _sections = []
+    _sections!.append([]);
+    
+    var section = 0
+    for idx in 0 ..< kNumberOfImages {
+      let name = String.init(format: "image_%d.jpg", idx)
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/ViewController.swift]: `init`

File: examples/CustomCollectionView-Swift/Sample/ViewController.swift [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+  
+  
+  func collectionNode(_ collectionNode: ASCollectionNode, nodeForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> ASCellNode {
+    let textAttributes : NSDictionary = [
+      NSFontAttributeName: UIFont.preferredFont(forTextStyle: UIFontTextStyle.headline),
+      NSForegroundColorAttributeName: UIColor.gray
+    ]
+    let textInsets = UIEdgeInsets.init(top: 11, left: 0, bottom: 11, right: 0)
+    let textCellNode = ASTextCellNode.init(attributes: textAttributes as! [AnyHashable : Any], insets: textInsets)
+    textCellNode.text = String.init(format: "Section %zd", indexPath.section + 1)
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/ViewController.swift]: You don't need the `.init(...)` you can just use `(...)`

File: examples/CustomCollectionView-Swift/Sample/ViewController.swift [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    self.view.addSubnode(_collectionNode!)
+  }
+  
+  override func viewWillLayoutSubviews() {
+    _collectionNode?.frame = self.view.bounds;
+  }
+  
+  func collectionNode(_ collectionNode: ASCollectionNode, nodeForItemAt indexPath: IndexPath) -> ASCellNode {
+    let image = _sections?[indexPath.section][indexPath.item]
+    return ImageCellNode.init(with: image!)
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/ViewController.swift]: You don't need the `.init(...)` you can just use `(...)`

File: examples/CustomCollectionView-Swift/Sample/ViewController.swift [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+      }
+    }
+    
+    let layout = MosaicCollectionViewLayout.init()
+    layout.numberOfColumns = 3;
+    layout.headerHeight = 44;
+    layout.delegate = self
+    _layoutInspector = MosaicCollectionViewLayoutInspector.init()
+    
+    _collectionNode = ASCollectionNode.init(frame: CGRect.zero, collectionViewLayout: layout)
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/ViewController.swift]: You don't need the `.init(...)` you can just use `(...)`

File: examples/CustomCollectionView-Swift/Sample/ViewController.swift [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+//
+
+import UIKit
+import AsyncDisplayKit
+
+class ViewController: UIViewController, MosaicCollectionViewLayoutDelegate, ASCollectionDataSource, ASCollectionDelegate {
+  
+  var _sections: [[UIImage]]?
+  var _collectionNode: ASCollectionNode?
+  var _layoutInspector: MosaicCollectionViewLayoutInspector?
```
Comment (george-gw) [inline examples/CustomCollectionView-Swift/Sample/ViewController.swift]: None of these need to be optionals. Initialize sections and layoutInspector here and the collectionNode before `super.init`.
The collectionNode and the layoutInspector will never change, so convert those two to `let` statements.

**Comment by rajeevwhiz** [comment]
Thanks @george-gw. Checked in all the changes suggested.

**Comment by hannahmbanana** [comment]
Thank you @rajeevwhiz for contributing and @george-gw for the swift style comments!

**Comment by george-gw** [comment]
Happy to help @rajeevwhiz @hannahmbanana :)

**Pull request merged**